sanic/server.py

```diff
@@ -888,15 +868,42 @@ def bind_socket(host: str, port: int, backlog=100) -> socket:
         sock = socket.socket()
     sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
     sock.bind((host, port))
+    sock.listen(backlog)
     return sock
 
 
-def remove_dead_socket(host: str) -> None:
-    """Remove dead unix socket during server exit, if host refers to one."""
-    if not host or not host.lower().startswith("unix:"):
-        return
-    path = host[5:]
-    if os.path.exists(path) and stat.S_ISSOCK(os.stat(path).st_mode):
+def bind_unix_socket(path: str, *, mode=0o666, backlog=100) -> socket:
+    """Create unix socket.
+    :param path: filesystem path
+    :param backlog: Maximum number of connections to queue
+    :return: socket.socket object
+    """
+    if os.path.exists(path) and not stat.S_ISSOCK(os.stat(path).st_mode):
+        raise FileExistsError(f"Existing file is not a socket: {path}")
+    sock = socket.socket(socket.AF_UNIX)
+    try:
+        # Atomic zero-downtime socket replace
+        tmp_path = f"{path}.{uuid4().hex[:8]}"
```

Comment:
Would `os.getpid()` be a simpler solution? It is already ensured to be unique identifier (at least inside a pid namespace (e.g. single container)). If the file would exists, it's would be from an already dead process.

Reply:
Stale sockets might still be left from before reboot, causing services fail to start. There are apparently also some security concerns, in particular that if another user creates a symlink in place of the socket, bind on MacOS follows that symlink and creates a socket at symlink's target. Eight random hex digits is far less likely to be hit by these issues than a pid, although it might be necessary to use even more.

A minor issue with this approach is that the initial random name sticks, and is returned as sockname on the server side and as peername on the clients.